storage/replica.go

```diff
@@ -1072,6 +1076,51 @@ func (r *Replica) maybeGossipConfigsLocked(match func(configPrefix proto.Key) bo
 	}
 }
 
+// maybeGossipSystemMap scans the entire SystemDB span and gossips it.
+// The first call is on NewReplica. Further calls come from the trigger
+// on an EndTransactionRequest.
+//
+// Note that maybeGossipSystemMap gossips information only when the
+// lease is actually held. The method does not request a leader lease
+// here since LeaderLease and applyRaftCommand call the method and we
+// need to avoid deadlocking in redirectOnOrObtainLeaderLease.
+// TODO(tschottdorf): Can possibly simplify.
+func (r *Replica) maybeGossipSystemMap() {
+	r.Lock()
+	defer r.Unlock()
+	r.maybeGossipSystemMapLocked()
+}
+
+func (r *Replica) maybeGossipSystemMapLocked() {
+	if r.rm.Gossip() == nil || !r.isInitialized() {
+		return
+	}
+
+	if lease := r.getLease(); !lease.OwnedBy(r.rm.RaftNodeID()) || !lease.Covers(r.rm.Clock().Now()) {
+		// Do not gossip when a leader lease is not held.
+		return
+	}
+
+	ctx := r.context()
+	// TODO(marc): check for bad split in the middle of the SystemDB span.
+	systemMap, hash, err := loadSystemMap(r.rm.Engine())
+	if err != nil {
+		log.Errorc(ctx, "could not load SystemDB map: %s", err)
+		return
+	}
+	if bytes.Equal(r.systemDBHash, hash) {
+		return
+	}
+
+	r.systemDBHash = hash
+	if log.V(1) {
+		log.Infoc(ctx, "gossiping SystemDB map from store %d, range %d", r.rm.StoreID(), r.Desc().RangeID)
+	}
+	if err := r.rm.Gossip().AddInfoProto(gossip.KeySystemDB, systemMap, 0); err != nil {
```

Comment:
Why does this gossip the whole thing as one key? Is there any reason not to gossip each kv key as a separate gossip key? I suppose needing tombstones is one reason, but the TTL should suffice for now.

The other problem here is that this is gossiped with a TTL of zero. Why is that OK?